design/cassandra/design.md

```diff
@@ -0,0 +1,109 @@
+# Kubernetes Operator for Apache Cassandra
+
+## Why ?
+
+* Cassandra is one of the most popular NoSQL databases.
+* Peer-to-Peer write-anywhere design for very high write and read throughput.
+* Despite those advantages, maintaining a Cassandra cluster is a highly manual and tiring task, which requires extensive familiarity with Cassandra's internals.
+* An operator will automate all the manual interventions that a human administrator normally has to perform and provide a production-ready database out-of-the-box.
+
+## Features
+
+The operator will provide all the important features of Cassandra while exposing a user-friendly declarative UI, in accordance to Kubernetes principles.
+More specifically, those features include:
+
+1. Cluster Creation
+2. (Auto)Scaling 
+3. Failed Cassandra Node Replacement
+4. Schedule Node Repair
+5. Backup/Restore
+6. Metrics endpoint for Kubernetes
+
+## Design
+
+### Overview
+
+* This operator will use the pattern: `CRD` + `Controller` + `Sidecar`
+
+![operator-design-overview](media/operator_overview.jpg)
+
+### Cassandra-Kubernetes Mapping
+
+* The design of the Cassandra CRD will follow the Cassandra terminology as much as possible, in order to be familiar to existing users of Cassandra and not confuse new users studying the Cassandra docs.
+
+* Cassandra abstracts resources in the following order:
+`Cluster` -> `Datacenter` -> `Rack` -> `Node`
+
+* We map those abstractions to Kubernetes in the following way:
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
+|     Node      |     Pod       |
+
+An example CRD would look something like this: 
+
+``` yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  version: "0.1"
+  dataCenters:
+  - name: "dc-1"
+    racks:
+    - name: "rack-1"
+      replicas: 3
+      nodeSelector:  
+        dc: 1
+        rack: 1
+      storage:
+        size: "500Gi"
+        storageClassName: "fast"
+      resources:
+        requests:
+          cpu: "2000m"
+          memory: "4Gi"
+        limits:
+          cpu: "2000m"
+          memory: "4Gi"
+  backup:
+    storageType: "S3"
+    schedule: "*/30 * * * *"
+  image:
```

Comment:
Is it necessary to specify the images here?
This can be taken care of by the Cassandra operator code.

Reply:
I think it is useful to expose the Cassandra image version, so that the user can have the freedom of choice. My goal is for the operator to use the [official](https://hub.docker.com/_/cassandra/) Cassandra images with a custom entrypoint.
I would agree that we could make this optional and resort to a reasonable default (eg the latest tag for each image).